멘티 일정 조회 및 인증 객체 Annotation 구현

itsjustsora opened the pull request:
## 내용
멘티 일정 조회 기능을 구현했습니다.
- 인증된 회원 정보와 날짜 정보로 Coaching schedule 조회
- 인증된 회원 정보로 Coaching relationship 조회

itsjustsora committed:
[#23] :sparkles: 인증 객체 커스텀 어노테이션 추가

itsjustsora committed:
[#23] :sparkles: Coaching relationship 추가
- 도메인 및 Response 객체 추가
- Mapper 추가

itsjustsora committed:
[#23] :sparkles: Coaching schedule 추가
- 도메인 및 Response 객체 추가
- Mapper 추가

itsjustsora committed:
[#23] :sparkles: 멘티 일정 조회 기능 추가

f-lab-kyle commented on src/main/java/com/flab/comen/global/auth/PrincipalDetails.java:
```diff
@@ -11,7 +11,7 @@
 public class PrincipalDetails implements UserDetails {
 
-	private Member member;
+	public Member member;
```
바꾸신 이유가 있으실까요~?

itsjustsora replied:
인증된 객체를 가져올 때 Member 객체를 반환하도록 했는데 private일 때 public이 아니어서 값을 가져오지 못한다는 에러 메시지가 반환되어서 public으로 변경했습니다!

f-lab-kyle commented on src/main/java/com/flab/comen/mentee/controller/MenteeController.java:
```diff
@@ -0,0 +1,37 @@
+		this.menteeService = menteeService;
+	}
+
+	@GetMapping(value = {"", "{date}"})
```
좀 고민이 되네요. `GET /api/v1/mentees` 이라면 멘티의 정보를 조회할것만 같아서요~
스케쥴 관련된 path가 있다면 더 명시적이지 않을까 싶네요.

itsjustsora replied:
제가 계속 UI 기준으로 생각하다 보니 기본 조회를 일정 조회라고 생각해서 명시적인 걸 생각을 못한 것 같습니다..!
조금 더 확실하게 의도를 나타낼 수 있도록 수정해 보겠습니다.

f-lab-kyle commented on src/main/java/com/flab/comen/mentee/service/MenteeService.java:
```diff
@@ -0,0 +1,45 @@
+		this.coachingRelationshipMapper = coachingRelationshipMapper;
+	}
+
+	public Optional<ScheduleResponse> getSchedule(Member member, LocalDate date) {
```
Request 전부를 들고오는 느낌인데요~ 사용할 필드만 넘기는게 좋아보입니다

itsjustsora replied:
인증된 객체의 id만 전달하도록 수정하겠습니다.

f-lab-kyle commented on src/main/java/com/flab/comen/mentee/service/MenteeService.java:
```diff
@@ -0,0 +1,45 @@
+	}
+
+	public Optional<ScheduleResponse> getSchedule(Member member, LocalDate date) {
+		if (date == null) {
+			date = LocalDate.now();
+		}
```
기본값 처리는 내부에서가 아닌 외부에서 주입하는게 어떨까요~!?

itsjustsora replied:
기본값 처리는 외부에서 하도록 변경하겠습니다!

f-lab-kyle commented on src/main/java/com/flab/comen/relationship/domain/CoachingRelationship.java:
```diff
@@ -0,0 +1,40 @@
+		this.expiredDt = expiredDt;
+	}
+
+	public static CoachingRelationship of(Long coachId, Long menteeId, boolean activeYn, LocalDate startedDt,
+		LocalDate expiredDt) {
+		return CoachingRelationship.builder()
+			.coachId(coachId)
+			.menteeId(menteeId)
+			.activeYn(activeYn)
+			.startedDt(startedDt)
+			.expiredDt(expiredDt)
+			.build();
+	}
```
혹시 빌더를 사용하지 않고 of를 재 정의한 이유가 있을까요?

itsjustsora replied:
추후 기능에 따라 생성자가 추가되더라도 기능을 명시적으로 확인할 수 있게 정적 팩토리 메서드를 사용하려고 
기본적인 매개변수를 받는 생성자도 정적 팩토리 메서드로 만들어 보았습니다!
new 생성자로 만들어서 필드 별로 값을 지정하는 것보다 builder가 가독성이 있는 것 같아 
매개변수 생성자에 builder 어노테이션을 사용했는데 생각해 보니 중복되는 생성자를 만든 것 같습니다 🥲

f-lab-kyle commented on src/main/java/com/flab/comen/relationship/dto/RelationshipResponse.java:
```diff
@@ -0,0 +1,25 @@
+	LocalDate startedDt,
+	LocalDate expiredDt
+) {
+	public static RelationshipResponse from(CoachingRelationship coachingRelationship) {
+		return new RelationshipResponse(
+			coachingRelationship.getTid(),
+			coachingRelationship.getCoachId(),
+			coachingRelationship.getMenteeId(),
+			coachingRelationship.isActiveYn(),
+			coachingRelationship.getStartedDt(),
+			coachingRelationship.getExpiredDt()
+		);
+	}
```
요런 형 변환 (Entity to DTO같은) 메소드는 더욱더 테스트 코드가 들어가야 할것 같아요. 필드 하나씩 빼먹는 경우가 있더라고요 ㅎ

itsjustsora replied:
넵 알겠습니다!

f-lab-kyle commented on src/main/resources/mapper/relationship/CoachingRelationshipMapper.xml:
```diff
@@ -0,0 +1,17 @@
+
+    <select id="findByMenteeIdAndActiveYn" parameterType="Long"
+            resultType="com.flab.comen.relationship.domain.CoachingRelationship">
+        SELECT *
```
가급적 `*` 보다는 명시적으로 필드명을 작성하는게 좋을것 같아요.
테이블 컬럼이 변경됨에따라 예상치 못한 매핑이 틀어지는 경우를 방지하기 위해서요.

itsjustsora replied:
필드명을 작성하는 걸로 수정하겠습니다.

f-lab-kyle commented on src/main/resources/mapper/relationship/CoachingRelationshipMapper.xml:
```diff
@@ -0,0 +1,17 @@
+        SELECT *
+        FROM coaching_relationship
+        WHERE mentee_id = #{menteeId}
+          AND active_yn = 'Y'
```
```suggestion
          AND active_yn = #{activeYn}
``` 

하드코딩보다 이처럼 주입받아서 처리해두면 활용하기에 더 좋을것 같아요!

itsjustsora replied:
최대한 외부에서 주입받아서 사용할 수 있도록 코드 변경하겠습니다!

f-lab-kyle commented on src/main/resources/mapper/relationship/CoachingRelationshipMapper.xml:
```diff
@@ -0,0 +1,17 @@
+
+
+    <select id="findByMenteeIdAndActiveYn" parameterType="Long"
+            resultType="com.flab.comen.relationship.domain.CoachingRelationship">
```
https://mybatis.org/mybatis-3/ko/configuration.html#typealiases 요런걸 활용하면 깔끔해질수도?

itsjustsora replied:
훨씬 깔끔하게 변경할 수 있을 것 같습니다! 감사합니다 🙂

f-lab-kyle commented on src/main/resources/mapper/schedule/CoachingScheduleMapper.xml:
```diff
@@ -0,0 +1,17 @@
+        SELECT *
+        FROM coaching_schedule
+        WHERE mentee_id = #{menteeId}
+          AND possible_dt LIKE #{date} '%'
```
이 쿼리에서 like 검색을 해야하는 이유를 잘 모르겠어요

itsjustsora replied:
possible_dt 컬럼에서는 코칭이 가능한 날짜와 시간 정보를 모두 가지고 있는데 
일정 조회에서는 날짜를 기준으로만 조회하다 보니 LIKE 연산자를 이용해서 날짜 정보를 기준으로 조회하려고 했습니다!

f-lab-kyle reviewed (commented):
범위가 길어진다면 default branch를 기준으로 feature-a branch를 생성하고 거기서 다시 feature-a1, feature-a2 이런식으로 생성한 다음에 feature-a 으로 pullrequest를 작성하는 방법도 있을것 같아요~
고생하셨습니당~

itsjustsora committed:
[#23] :sparkles: path 변경 및 기존 코드 개선
- schedule path 추가
- 조회 조건들에 사용되는 파라미터는 외부에서 주입되도록 변경
- alias-package 설정 사용하여 resultType 개선
- select문에서 * 제거 후 필드명 선언
- builder 제거
